### PR TITLE
JavaScript RegEx Validation

### PR DESCRIPTION
**Pull request by codysmith287**
This is an initial draft of a JavaScript RegEx validation readme. Let me know your thoughts at your convenient. No time sensitively. Thanks.

### CODE REVIEW

File: code-quality/validation/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+```javascript
+/([!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~])+/g
+```
+
+## RegEx Email Validation Example (JavaScript)
+
+- An email is a string (a subset of ASCII characters) separated into two parts by @ symbol, a "personal_info" and a domain, that is personal_info@domain. The length of the personal_info part may be up to 64 characters long and domain name may be up to 253 characters.
+
+```javascript
+/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/
```
Comment (carlosvargas) [inline code-quality/validation/README.md]: A few things:

* TLDs can be more than 3 characters now: https://en.wikipedia.org/wiki/List_of_Internet_top-level_domains#ICANN-era_generic_top-level_domains.
* I would make the groups non-capturing groups `(?:[\.-]?\w+)*` just to prevent someone accidentally thinking that they could grab user/domain/etc from the regex.

Comment (codysmith287) [reply code-quality/validation/README.md]: You're right, thanks for the feedback. I'll change to up to 24 characters as that's the longest TDL currently in existence. The longest in English is 14 characters (.cancerresearch).

I'll also change to the groups to non-capturing groups.

**Comment by DropsOfSerenity** [comment]
Can we consider putting a note here that email regexp's are not preferred. The best pattern for validating email addresses is to make accounts confirmable, e.g send an email to the user signing up and have them click a link to confirm that account. This filters against incorrect and bad email addresses.

Thoughts?

**Comment by codysmith287** [comment]
@DropsOfSerenity Great point. Like @carlosvargas said, I think we should do both and it helps the UX. I can add a note that states regex should not be the only means for email validation in your application and that accounts should be confirmable whenever possible.

File: code-quality/validation/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+<!-- Last updated 9-25-19 -->
+
+# Input Validation 
+
+Input validation is is your first line of defense when creating a secure application.
```
Comment (stephengtuggy) [inline code-quality/validation/README.md]: Omit second "is"

File: code-quality/validation/README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+- Check for at least one numeric character: 
+```javascript
+/([0-9]+)/g
+```
+
+
+- Check for at least one special character:
+```javascript
+/([!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~])+/g
```
Comment (stephengtuggy) [inline code-quality/validation/README.md]: I don't think we should include -, <, or ', as these can be used in SQL Injection and similar attacks.

File: code-quality/validation/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+## RegEx Email Validation Example (JavaScript)
+
+- An email is a string (a subset of ASCII characters) separated into two parts by @ symbol, a "personal_info" and a domain, that is personal_info@domain. The length of the personal_info part may be up to 64 characters long and domain name may be up to 253 characters.
+
+```javascript
+/^\w+(?:[\.-]?\w+)*@\w+(?:[\.-]?\w+)*(?:\.\w{2,24})+$/
+```
+
+- Fun fact: The longest Top Level domain in English is .cancerresearch which 14 characters. The longest currently in existance is 24 characters.
```
Comment (stephengtuggy) [inline code-quality/validation/README.md]: "which is"

Comment (stephengtuggy) [reply code-quality/validation/README.md]: "existence"

File: code-quality/validation/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+## Whitelist or Blacklist? 
+
+The first concept of good input validation is whitelisting versus blacklisting. Whitelist, or inclusive validation, defines a set of valid characters while blacklist, or exclusive validation, defines a set of invalid characters to try to remove. 
+
+Input validation using blacklisting is generally more challenging to implement so it is better practice to perform validation using whitelisting. A great way of defining a whitelist for input validation is to leverage Regular Expressions(RegEx). 
+
+## RegEx Password Validation Examples (JavaScript)
+
+- Mininum length of any 6 chars but newline:
```
[the remark this answers is not in the document]
Comment (codysmith287) [reply code-quality/validation/README.md]: I agree eight characters should be the standard so I'll go ahead and change it to eight.

File: code-quality/validation/README.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+# Input Validation 
+
+Input validation is is your first line of defense when creating a secure application.  
+
+## Whitelist or Blacklist? 
+
+The first concept of good input validation is whitelisting versus blacklisting. Whitelist, or inclusive validation, defines a set of valid characters while blacklist, or exclusive validation, defines a set of invalid characters to try to remove. 
+
+Input validation using blacklisting is generally more challenging to implement so it is better practice to perform validation using whitelisting. A great way of defining a whitelist for input validation is to leverage Regular Expressions(RegEx).
```
[the remark this answers is not in the document]
Comment (codysmith287) [reply code-quality/validation/README.md]: I browsed a few different sources, but the general consent seems to be whitelist are better for security. 

https://choosetoencrypt.com/tech/whitelist-vs-blacklist/

Specifically the bottom paragraph of this article. 

"In general though, if you are trying to make your system as secure as possible the best choice will be a whitelist. Whitelisting is the choice for organizations with high-security standards or sensitive information. Regardless of the extra time required to maintain an up-to-date whitelist."

File: code-quality/validation/README.md (hunk 2 above)
Comment (codysmith287) [reply code-quality/validation/README.md]: Good catch.

File: code-quality/validation/README.md (hunk 4 above)
Comment (codysmith287) [reply code-quality/validation/README.md]: Thanks, good eye.

File: code-quality/validation/README.md (hunk 3 above)
Comment (carlosvargas) [reply code-quality/validation/README.md]: We should be sanitizing/encoding the data we get in the server, so the SQL injection shouldn't be a thing.

We also shouldn't prevent the user from using them if it makes the password stronger.

File: code-quality/validation/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (carlosvargas) [reply code-quality/validation/README.md]: It's also mentioned in the OWASP guide (which is the de facto standard used for a lot of security certifications):
 https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#whitelisting-vs-blacklisting

**Comment by DropsOfSerenity** [comment]
> @DropsOfSerenity Great point. Like @carlosvargas said, I think we should do both and it helps the UX. I can add a note that states regex should not be the only means for email validation in your application and that accounts should be confirmable whenever possible.

https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#email-validation-basics

File: code-quality/validation/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (codysmith287) [reply code-quality/validation/README.md]: Awesome, thank you for pointing out this resource, @carlosvargas!

**Comment by carlosvargas** [comment]
@stephengtuggy can you re-review this PR please?

File: code-quality/validation/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+## RegEx Email Validation Example (JavaScript)
+
+- DISCLAIMER: RegEx email validation should not be your sole source of email validation, rather it should be used to help your UI display an `invalid email` message. Whenever possible, one should include a process to confirm a given email address (e.g. send an email for the user confirm their email address).
+
+- An email is a string (a subset of ASCII characters) separated into two parts by @ symbol, a "personal_info" and a domain, that is personal_info@domain. The length of the personal_info part may be up to 64 characters long and domain name may be up to 253 characters.
+
+```javascript
+/^\w+(?:[\.-]?\w+)*@\w+(?:[\.-]?\w+)*(?:\.\w{2,24})+$/
```
Comment (DropsOfSerenity) [inline code-quality/validation/README.md]: Why put a time sensitive limit on this when we know it will be out of date in a few years? Suggest `{2,}` instead to remove the upper bound, or find the max upper bound allowed by the spec, vs the max currently in use.

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: This regex also does not allow for tags in email address which are *valid* such as `user+tag@mumble.dom` rather than trying to account for all the possible variants, i propose something akin to  [[anything]]@[[anything]].[[anything]] essentially, try to make this as permissive as possible, the minimum validation required to help the user from doing something stupid, all else should be handled by confirmation email.

Comment (codysmith287) [reply code-quality/validation/README.md]: @DropsOfSerenity fair enough, I'll remove the upper bound for the TDL and change the regex to be more permissive.

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: @codysmith287 awesome! :) This way we prevent false negatives, it can be frustrating for people with unusual email addresses or strange requirements we can't foresee (like tags) to be denied sign up for a website with a perfectly valid email address.

Comment (codysmith287) [reply code-quality/validation/README.md]: @DropsOfSerenity Yeah, I agree. Thanks for the feedback! I went ahead and made those changes to email regex along with the notes below. I also added a warning right above it stating proper care should be taken to sanitize the input to prevent any SQL injection attacks.

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: Stilll have some issues with this regex, prefer none at all or `/.+@.+\..+/`

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: or if you like yours better, at the very least, remove the `+` after the `@`

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: Also the language around sql injection is out of place being called out specifically on email, frontend validation will never stop people who want to use sql injection from the frontend, they will post the data directly where the javascript validators won't catch any of it regardless of what input you are using. That is more a backend concern, most ORM's handle it form you, I think that belongs on a backend security article and not here.

Comment (DropsOfSerenity) [reply code-quality/validation/README.md]: Other than all these nitpicks, I like the rest of this PR.

**Comment by DropsOfSerenity** [comment]
> The email regex is getting there, but still has at least one issue, I believe. It will match any number of at signs (`@`) in the middle of the email address. E.g., `bla@@@@@gmail.com` matches.
> 
> Also, "TDL" should be changed to "TLD", in at least two places.

image

Copying this here since i think i put it on an outdated code snippet and it might not get seen.

**Comment by codysmith287** [comment]
@stephengtuggy Good catch, thank you.

**Comment by codysmith287** [comment]
@stephengtuggy I see what you're saying about the bla@@@@gmail.com, but to Justin's point, the email should be confirmed.

**Comment by codysmith287** [comment]
@DropsOfSerenity I went ahead and removed the SQL injection warning as I agree it can be covered in a backend security article. I changed my regex example to even more permissive. It essentially allows anything except whitespaces in the personal_info, domain name, and TLD. I kept the non-capturing groups per some earlier changes requested from Carlos. I would still argue that the TLD should be at least two characters long. I included your example `.+@.+\..+` as an alternative code example.

**Comment by coreyshuman** [comment]
@codysmith287 did you see this issue? https://github.com/Shift3/standards-and-practices/issues/39#issuecomment-408524585
I spent some time a while back working on options for email validation. Let me know if you think there's any value in the version defined in this comment. (the whole thread is worth a read as well).
#39

**Comment by coreyshuman** [comment]
@carlosvargas @mwallert @ryekerjh @DropsOfSerenity If we're going to start defining standard RegEx snippets (which I like), I think we should have associated unit tests for each one as well. A good use case would be the conversation regarding the email regex above. We should be adding these extra edge cases to a test set to run against the proposed standard snippet.

Definitely doesn't need to be part of this initial submission, but wanted to see what you guys thought of the idea in general. I feel like it would be a very effective way to build confidence behind the snippets we use.

**Comment by mwallert** [comment]
@coreyshuman I agree, @ryekerjh suggested that that any code snippet we add should come with a test. Should we move for this to be: _before this is approved and merged_ it needs to come with a test?

**Comment by DropsOfSerenity** [comment]
@coreyshuman I would normally agree, but what i'm trying to make clear here is that complicated email regex is not the preferred pattern for signup  or email validation anyway. Attempting to send an email to the address specified is. Provided a permissive regex, or none at all (or just asking the user to enter their email twice) while sending a confirmation email, is a 100% method to ensure you end up with a valid email address, and 100% method to make sure you have no false negatives.

When you run into an "edge case" in your complicated regular expression you have to do the follow -> find the fix, hope you don't implement a regression possibly in other untested parts of the regex -> backport to all running applications using the old regex -> make sure all old versions of applications are updated -> etc. etc. etc.

I think that have an email regex may be valuable for things other than sign up fields, but I want it to be clear that in my opinion for sign in/sign up this is not the preferred pattern of validation, nor does it enhance security.

**Comment by DropsOfSerenity** [comment]
@coreyshuman As an example (and where it looks like the source of your comment was) you mention here https://github.com/Shift3/row-forge-ns/issues/796 that a malformed email got through your sign up process. Your solution was to implement a regex pattern to sign up, this fixes the malformed problem but this does not ensure that the user has entered the correct email address. They could enter an email address that is "well formed" but not "correct" e.g typo their own email address. Confirmation emails are the widely accepted standard on sign up for solving this problem and would nip both the malformed issue and an issue that wasn't solved on row-forge by just implementing regex.

**Comment by stephengtuggy** [comment]
@DropsOfSerenity I agree that ultimately, we should use confirmation emails for validation. But it's quite a bit more convenient for the user if we catch obvious formatting errors as early as possible, i.e. in the browser client.

I think either @coreyshuman 's regex from #39 , or @codysmith287 's latest regex here, would work for this purpose. Then we don't need to spend any more time on it.

File: code-quality/validation/README.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+- Check for at least one special character:
+```javascript
+/([!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~])+/g
+```
+
+## RegEx Email Validation Example (JavaScript)
+
+- DISCLAIMER: RegEx email validation should not be your sole source of email validation, rather it should be used to help your UI display an `invalid email` message. Whenever possible, one should include a process to confirm a given email address (e.g. send an email for the user confirm their email address).
```
Comment (DropsOfSerenity) [inline code-quality/validation/README.md]: Would appreciate further discussion on this topic, permissive vs non-permissive regex here, permissive regex makes sure we don't run into false negatives (e.g people with a correct email being unable to sign up for the systems we build) I would like to hear advocates for non-permissive regex's for sign ups defend their position on this point a little more.

Otherwise I would prefer this to say that on sign ups, we prefer permissive regex if any, and the primary source of validation and verification of correct email address to be verification emails (a standard used across the industry)

**Comment by Karvel** [comment]
I added URL RegEx validation that follows best practices to a project. The RegEx and tests came from here (there are excellent conversations in the comments also): https://gist.github.com/dperini/729294

This RegEx requires a protocol, which is necessary for opening a URL in a new tab/window.

The implemented RegEx and unit tests are here (I removed `ftp` as an allowed protocol for my use case):

https://github.com/Shift3/sinclair-scanner-website/blob/development/src/app/infrastructure/utils/constants.ts#L10

https://github.com/Shift3/sinclair-scanner-website/blob/development/src/app/infrastructure/validation/url-validation.spec.ts

This could be overkill in certain cases, but I believe that its false positives and false negatives are reasonable.

Here are the raw snippets for convenience:

```ts
const urlRegex = new RegExp(
	'^' +
	// protocol identifier (optional)
	// short syntax // still required
	'(?:(?:(?:https?):)?\\/\\/)' +
	// user:pass BasicAuth (optional)
	'(?:\\S+(?::\\S*)?@)?' +
	'(?:' +
	// IP address exclusion
	// private & local networks
	'(?!(?:10|127)(?:\\.\\d{1,3}){3})' +
	'(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})' +
	'(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})' +
	// IP address dotted notation octets
	// excludes loopback network 0.0.0.0
	// excludes reserved space >= 224.0.0.0
	// excludes network & broadcast addresses
	// (first & last IP address of each class)
	'(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])' +
	'(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}' +
	'(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))' +
	'|' +
	// host & domain names, may end with dot
	// can be replaced by a shortest alternative
	// (?![-_])(?:[-\\w\\u00a1-\\uffff]{0,63}[^-_]\\.)+
	'(?:' +
	'(?:' +
	'[a-z0-9\\u00a1-\\uffff]' +
	'[a-z0-9\\u00a1-\\uffff_-]{0,62}' +
	')?' +
	'[a-z0-9\\u00a1-\\uffff]\\.' +
	')+' +
	// TLD identifier name, may end with dot
	'(?:[a-z\\u00a1-\\uffff]{2,}\\.?)' +
	')' +
	// port number (optional)
	'(?::\\d{2,5})?' +
	// resource path (optional)
	'(?:[/?#]\\S*)?' +
	'$', 'i',
);```

```ts
!environment.testUnit
	? Logger.log('Unit skipped')
	: describe('[Unit] URLValidation validURL() Required', () => {
		const urlValidator = URLValidation.validURL(true);
		const urlControl = new FormControl('');
		const matchingValues: string[] = [
			'https://✪df.ws/123',
			'http://✪df.ws/123',
			'http://userid:password@example.com:8080',
			'http://userid:password@example.com:8080/',
			'http://userid@example.com',
			'http://userid@example.com/',
			'http://userid@example.com:8080',
			'http://userid@example.com:8080/',
			'http://userid:password@example.com',
			'http://userid:password@example.com/',
			'http://142.42.1.1/',
			'http://142.42.1.1:8080/',
			'http://➡.ws/䨹',
			'http://⌘.ws',
			'http://⌘.ws/',
			'http://foo.com/blah_(wikipedia)#cite-1',
			'http://foo.com/blah_(wikipedia)_blah#cite-1',
			'http://foo.com/unicode_(✪)_in_parens',
			'http://foo.com/(something)?after=parens',
			'http://☺.damowmow.com/',
			'http://code.google.com/events/#&product=browser',
			'http://j.mp',
			'http://foo.bar/?q=Test%20URL-encoded%20stuff',
			'http://مثال.إختبار',
			'http://例子.测试',
		];
		const failingValues: string[] = [
			'http://',
			'http://.',
			'http://..',
			'http://../',
			'http://?',
			'http://??',
			'http://??/',
			'http://#',
			'http://##',
			'http://##/',
			'http://foo.bar?q=Spaces should be encoded',
			'//',
			'//a',
			'///a',
			'///',
			'http:///a',
			'foo.com',
			'rdar://1234',
			'h://test',
			'http:// shouldfail.com',
			':// should fail',
			'http://foo.bar/foo(bar)baz quux',
			'ftps://foo.bar/',
			'http://-error-.invalid/',
			'http://-a.b.co',
			'http://a.b-.co',
			'http://0.0.0.0',
			'http://10.1.1.0',
			'http://10.1.1.255',
			'http://224.1.1.1',
			'http://1.1.1.1.1',
			'http://123.123.123',
			'http://3628126748',
			'http://.www.foo.bar/',
			'http://.www.foo.bar./',
			'http://10.1.1.1',
			'http://10.1.1.254',
		];

		it(`should return null if value matches a list of values that should work`, () => {
			matchingValues.forEach(value => {
				urlControl.setValue(value);
				expect(urlValidator(urlControl)).toEqual(null);
			});
		});

		it(`should return { invalidURL: 'Please enter a valid URL.' } if value matches a list of values that should fail`, () => {
			failingValues.forEach(value => {
				urlControl.setValue(value);
				const expectedValue = { invalidURL: 'Please enter a valid URL.' };
				expect(urlValidator(urlControl)).toEqual(expectedValue);
			});
		});
	});
```

File: code-quality/validation/README.md (hunk 8 above)
Comment (Karvel) [reply code-quality/validation/README.md]: I agree with this, but client validation provides nice UX. It doesn't prevent someone making a typo that is a well-formed email, but it does help with a typo that makes an invalid email.

**Comment by carlosvargas** [comment]
@codysmith287 can you please make any changes based on the latest comment?

**Comment by ryekerjh** [comment]
@Shift3/pr-team I am closing this PR and opening a new one that merges this reviewed file with the new file structure. I will be approving the PR as it has been reviewed and approved.
